pkg/vpcmodel/groupingSelfLoop.go

```diff
@@ -0,0 +1,333 @@
+						}
+						keyToMergeCandidates[mergeCandidateKey][key] = struct{}{}
+					}
+				}
+			}
+		}
+	}
+	return keyToMergeCandidates
+}
+
+func (g *GroupConnLines) addVsiSubnetToBucket(ep EndpointElem, bucket string) string {
+	if isVsi, node := isEpVsi(ep); isVsi {
+		// if ep is groupedEndpointsElems of vsis then all belong to the same subnet
+		return bucket + ";" + g.c.getSubnetOfNode(node).Name()
+	}
+	return bucket
+}
+
+// returns a pair <bool, node>:
+// if the endpoint element represents a vsi or is a slice of elements the first of which represents vsi
+// then it returns <true, the vsi or the first vsi>
+// otherwise it returns <false, nil>
+func isEpVsi(ep EndpointElem) (bool, Node) {
+	if _, ok := ep.(*groupedEndpointsElems); ok {
+		ep1GroupedEps := ep.(*groupedEndpointsElems)
+		if node, ok := (*ep1GroupedEps)[0].(Node); ok {
+			if node.IsInternal() {
+				return true, node
+			}
+		}
+		return false, nil
+	}
+	if node, ok := ep.(Node); ok {
+		if node.IsInternal() {
+			return true, node
+		}
+	}
+	return false, nil
+}
+
+// creates an aux database in which all the grouped endpoints are stored in a set
+func createGroupingSets(groupingSrcOrDst map[string][]*GroupedConnLine, srcGrouping bool) map[string]map[string]struct{} {
+	keyToGroupedSets := make(map[string]map[string]struct{})
+	for key, groupedConnLine := range groupingSrcOrDst {
+		mySet := make(map[string]struct{})
+		for _, line := range groupedConnLine {
+			srcOrDst := line.getSrcOrDst(srcGrouping)
+			mySet[srcOrDst.Name()] = struct{}{}
+		}
+		keyToGroupedSets[key] = mySet
+	}
+	return keyToGroupedSets
+}
+
+// computes delta between group connection lines as defined in https://github.com/np-guard/vpc-network-config-analyzer/issues/98
+// and in the beginning of this file
+func deltaBetweenGroupedConnLines(srcGrouping bool, groupedConnLine1, groupedConnLine2 []*GroupedConnLine,
+	setToGroup1, setToGroup2 map[string]struct{}) bool {
+	// at least one of the keys must be a single vsi/subnet for the self loop check to be meaningful
+	if len(elemInKeys(srcGrouping, *groupedConnLine1[0])) > 1 && len(elemInKeys(srcGrouping, *groupedConnLine2[0])) > 1 {
+		return false
+	}
+	// is there is a real delta between sets and not only due to self loop
+	set1MinusSet2 := setMinusSet(srcGrouping, *groupedConnLine2[0], setToGroup1, setToGroup2)
+	set2MinusSet1 := setMinusSet(srcGrouping, *groupedConnLine1[0], setToGroup2, setToGroup1)
+	if len(set1MinusSet2) == 0 && len(set2MinusSet1) == 0 {
+		return true
+	}
+	return false
+}
+
+func elemInKeys(srcGrouping bool, groupedConnLine GroupedConnLine) []string {
+	srcOrDst := groupedConnLine.getSrcOrDst(srcGrouping)
+	return strings.Split(srcOrDst.Name(), commaSeparator)
+}
+
+func setMinusSet(srcGrouping bool, groupedConnLine GroupedConnLine, set1, set2 map[string]struct{}) map[string]struct{} {
+	minusResult := make(map[string]struct{})
+	for k := range set1 {
+		if _, ok := set2[k]; !ok {
+			minusResult[k] = struct{}{}
+		}
+	}
```

Comment:
A simpler type for representing a set can be `map[string]bool` instead of `map[string]struct{}` (when a key is added it is mapped to `true`) so that to check if `k` is in the set you simply use expression `if set[k] { ...} `  instead of 
```
if _, ok := set[k] ; ok{

}
```
consider to simplify in relevant places in the code

Reply:
I am familiar with this representation of sets; is it indeed simpler? because `Bool` can be `false` or `true`; as such, checking for existence in the map is not enough and the content of the map also needs to be checked. In `map[string]struct{}` it is clear that the value of the map is meaningless.

Reply:
Any value in the map is always mapped to `True`, so that the expression `set[k]` is True iff the key `k` is in the map

Reply:
I disagree here. I think the `Bool` in this context is confusing in the sense that one may assume that it can also be mapped to false, while `struct{}` clearly implies that the value is meaningless.

Reply:
ok, let's keep as is now